diff --git a/tools/PCKRetrievalTool/Enclave/Enclave.cpp b/tools/PCKRetrievalTool/Enclave/Enclave.cpp
index ae731cf..1aa70e5 100644
--- a/tools/PCKRetrievalTool/Enclave/Enclave.cpp
+++ b/tools/PCKRetrievalTool/Enclave/Enclave.cpp
@@ -38,6 +38,8 @@
 #include "sgx_report.h"
 #include "sgx_utils.h"
 
+#include "sgx_cpuid.h"
+
 
 uint32_t enclave_create_report(const sgx_target_info_t* p_qe3_target, sgx_report_t* p_report)
 {
diff --git a/tools/PCKRetrievalTool/Enclave/Enclave.edl b/tools/PCKRetrievalTool/Enclave/Enclave.edl
index 0a1210d..e698034 100644
--- a/tools/PCKRetrievalTool/Enclave/Enclave.edl
+++ b/tools/PCKRetrievalTool/Enclave/Enclave.edl
@@ -29,9 +29,8 @@
  *
  */
 
-
-
 enclave {
+    from "sgx_tstdc.edl" import sgx_oc_cpuidex;
     include "sgx_report.h"
 
     trusted {
diff --git a/tools/PCKRetrievalTool/Makefile b/tools/PCKRetrievalTool/Makefile
index 4596ee9..db99405 100644
--- a/tools/PCKRetrievalTool/Makefile
+++ b/tools/PCKRetrievalTool/Makefile
@@ -104,7 +104,7 @@ App_C_Flags := $(COMMON_FLAGS) -fPIC -Wno-attributes $(App_Include_Paths)
 
 App_Cpp_Flags := $(App_C_Flags) -std=c++11
 App_Link_Flags := $(SGX_COMMON_CFLAGS) -Wl,-z,relro,-z,now,-z,noexecstack
-App_Link_Flags +=  -lcurl -ldl -lpthread -Wl,-rpath=.
+App_Link_Flags +=  -lcurl -ldl -lpthread -Wl,-rpath=. -lsgx_urts
 
 
 App_Cpp_Objects := $(App_Cpp_Files:.cpp=.o)
@@ -187,7 +187,7 @@ endif
 ######## App Objects ########
 
 App/Enclave_u.c:
-	cd App && $(SGX_EDGER8R) --untrusted ../Enclave/Enclave.edl
+	cd App && $(SGX_EDGER8R) --untrusted ../Enclave/Enclave.edl --search-path $(SGX_SDK)/include
 	@echo "GEN  =>  $@"
 
 App/Enclave_u.o: App/Enclave_u.c
@@ -213,7 +213,7 @@ $(ENCLAVE_SIGNING_KEY):
 
 
 Enclave/Enclave_t.h: $(SGX_EDGER8R) Enclave/Enclave.edl
-	@cd Enclave && $(SGX_EDGER8R) --trusted ../Enclave/Enclave.edl 
+	@cd Enclave && $(SGX_EDGER8R) --trusted ../Enclave/Enclave.edl --search-path $(SGX_SDK)/include
 	@echo "GEN  =>  $@"
 
 Enclave/Enclave_t.c: Enclave/Enclave_t.h
