diff --git a/src/cbor.h b/src/cbor.h
index be5bbc7..edb160b 100644
--- a/src/cbor.h
+++ b/src/cbor.h
@@ -702,7 +702,7 @@ typedef CborError (*CborStreamFunction)(void *token, const char *fmt, ...)
 CBOR_API CborError cbor_value_to_pretty_stream(CborStreamFunction streamFunction, void *token, CborValue *value, int flags);
 
 /* The following API requires a hosted C implementation (uses FILE*) */
-#if !defined(__STDC_HOSTED__) || __STDC_HOSTED__-0 == 1
+#if !defined(TINYCBOR_FREESTANDING) && (!defined(__STDC_HOSTED__) || __STDC_HOSTED__-0 == 1)
 CBOR_API CborError cbor_value_to_pretty_advance_flags(FILE *out, CborValue *value, int flags);
 CBOR_API CborError cbor_value_to_pretty_advance(FILE *out, CborValue *value);
 CBOR_INLINE_API CborError cbor_value_to_pretty(FILE *out, const CborValue *value)
diff --git a/src/compilersupport_p.h b/src/compilersupport_p.h
index 0879801..a31f1ee 100644
--- a/src/compilersupport_p.h
+++ b/src/compilersupport_p.h
@@ -44,9 +44,9 @@
 #  include <stdbool.h>
 #endif
 
-#if __STDC_VERSION__ >= 201112L || (defined(__cplusplus) && __cplusplus >= 201103L) || (defined(__cpp_static_assert) && __cpp_static_assert >= 200410)
-#  define cbor_static_assert(x)         static_assert(x, #x)
-#elif !defined(__cplusplus) && defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 406) && (__STDC_VERSION__ > 199901L)
+//#if __STDC_VERSION__ >= 201112L || (defined(__cplusplus) && __cplusplus >= 201103L) || (defined(__cpp_static_assert) && __cpp_static_assert >= 200410)
+//#  define cbor_static_assert(x)         static_assert(x, #x)
+#if !defined(__cplusplus) && defined(__GNUC__) && (__GNUC__ * 100 + __GNUC_MINOR__ >= 406) && (__STDC_VERSION__ > 199901L)
 #  define cbor_static_assert(x)         _Static_assert(x, #x)
 #else
 #  define cbor_static_assert(x)         ((void)sizeof(char[2*!!(x) - 1]))
